kernel/src/engine/arrow_expression.rs

```diff
@@ -366,6 +364,193 @@ fn evaluate_expression(
     }
 }
 
+// Apply a schema to an array. The array _must_ be a `StructArray`. Returns a `RecordBatch where the
+// names of fields, nullable, and metadata in the struct have been transformed to match those in
+// schema specified by `schema`
+fn apply_schema(array: &dyn Array, schema: &DataType) -> DeltaResult<RecordBatch> {
+    let DataType::Struct(struct_schema) = schema else {
+        return Err(Error::generic(
+            "apply_schema at top-level must be passed a struct schema",
+        ))?;
+    };
+    let applied = apply_schema_to_struct(array, struct_schema)?;
+    Ok(applied.into())
+}
+
+// helper to transform an arrow field+col into the specified target type. If `rename` is specified
+// the field will be renamed to the contained `str`.
+fn transform_field_and_col(
+    arrow_field: &Arc<ArrowField>,
+    arrow_col: Arc<dyn Array>,
+    target_type: &DataType,
+    nullable: bool,
+    rename: Option<&str>,
+    metadata: Option<HashMap<String, String>>,
+) -> DeltaResult<(ArrowField, Arc<dyn Array>)> {
+    let transformed_col = apply_schema_to(&arrow_col, target_type)?.unwrap_or(arrow_col);
+    let transformed_field = arrow_field
+        .as_ref()
+        .clone()
```

Comment:
is this a non-Arc clone? that is, we actually deep clone and the make changes? interesting it isn't `mut`?

Reply:
> But at least [arrow] struct fields break the clone recursion by being Arc. Maybe we should consider doing that in kernel as well, so that schema manipulation doesn't deep-clone the whole tree every time?

Filed as https://github.com/delta-incubator/delta-kernel-rs/issues/404